pep-0697.rst

```diff
@@ -0,0 +1,357 @@
+If the instance struct ends with a variable-length array (such as
+in ``tuple`` or ``int``), subclasses cannot add their own extra data without
+detailed knowledge about how the superclass allocates and uses its memory.
+
+Some types, such as CPython's ``PyHeapType``, handle this by storing
+variable-sized data after the fixed-size struct.
+This means that any subclass can add its own fixed-size data.
+(Only one class in the inheritance hierarchy can use variable-sized data, though.)
+This PEP proposes API that makes this practice easier, and ensures the
+variable-sized data is properly aligned.
+
+Note that many variable-size types, like ``int`` or ``tuple``, do not use
+this mechanism. 
+This PEP does not propose any changes to existing variable-size types (like
+``int`` or ``tuple``) except ``PyHeapType``.
+
+
+Extending ``PyHeapType`` specifically
+-------------------------------------
+
+The motivating problem this PEP solves is creating metaclasses, that is,
+subclasses of ``type``.
+The underlying ``PyHeapTypeObject`` struct is both and variable-sized, and
+opaque in the limited API.
+
+Projects such as language bindings and frameworks that need to attach custom
+data to metaclasses currently resort to questionable workarounds.
+The situation is worse in projects that target the Limited API.
+
+For an example of the currently necessary workarounds, see:
+`nb_type_data_static <https://github.com/wjakob/nanobind/blob/f3044cf44763e105428e4e0cf8f42d951b9cc997/src/nb_type.cpp#L1085>`_
+in the not-yet-released limited-API branch of ``nanobind``
+(a spiritual successor of the popular C++ binding generator ``pybind11``).
+
+
+Rationale
+=========
+
+This PEP proposes a different model: instead of the superclass data being
+part of the subclass data, the extra space a subclass needs is specified
+and accessed separately.
+(How base class data is accessed is left to whomever implements the base class:
+they can for example  provide accessor functions, expose a part of its
+``struct`` for better performance, or do both.)
+
+The proposed mechanism allows using static, read-only ``PyType_Spec``
+even if the superclass struct is opaque, like ``PyTypeObject`` in
+the Limited API.
+
+Combined with a way to create class from ``PyType_Spec`` and a custom metaclass,
+this will allow libraries like nanobind or JPype to create metaclasses
+without making assumptions about ``PyTypeObject``'s memory layout.
+The approach generalizes to non-metaclass types as well.
+
+
+Specification
+=============
+
+In the code blocks below, only function headers are part of the specification.
+Other code (the size/offset calculations) are details of the initial CPython
+implementation, and subject to change.
+
+Relative ``basicsize``
+----------------------
+
+The ``basicsize`` member of ``PyType_Spec`` will be allowed to be zero or
+negative.
+In that case, it will specify the inverse of *extra* storage space instances of
```

Comment:
"inverse" makes me think of 1/x, not -x. How about this?
```suggestion
In that case, its absolute value will specify the amount of *extra* storage space instances of
```

Reply:
Thanks for the clarification!
(English math terminology is confusing. I looked up that [inverse can mean both](https://en.wikipedia.org/wiki/Inverse), and hoped the meaning's clear when dealing with integers and negatives.)

Reply:
Additive inverse is the correct terminology here but absolute value works too :)